Include Optimism (#345)
subquery/documentation: docs/quickstart/quickstart_chains/optimism.md

- [BlockHanders](../../build/manifest/optimism.md#mapping-handlers-and-filters): On each and every block, run a mapping function
- [TransactionHandlers](../../build/manifest/optimism.md#mapping-handlers-and-filters): On each and every transaction that matches optional filter criteria, run a mapping function
- [LogHanders](../../build/manifest/optimism.md#mapping-handlers-and-filters): On each and every log that matches optional filter criteria, run a mapping function

Note that the manifest file has already been set up correctly and doesn’t require significant changes, but you need to import the correct contract definitions and update the datasource handlers.

As we are indexing all claim events from the Optimism airdrop contract, the first step is to import the contract abi definition which can be obtained from [here](https://optimistic.etherscan.io/address/0xfedfaf1a10335448b7fa0268f56d2b44dbd357de#code). Copy the entire contract ABI and save it as a file called `airdrop.abi.json` in the `/abis` directory.

**Update the `datasources` section as follows:**

```yaml
dataSources:
  - kind: ethereum/Runtime # We use ethereum runtime since Optimism is a layer-2 that is compatible

>     startBlock: 100316590

Reviewer: ```suggestion
```

>     # startBlock: 9277162 # When the airdrop contract was deployed https://optimistic.etherscan.io/tx/0xdd10f016092f1584912a23e544a29a638610bdd6cb42a3e8b13030fd78334eba

Reviewer: ```suggestion
    startBlock: 9277162 # When the airdrop contract was deployed https://optimistic.etherscan.io/tx/0xdd10f016092f1584912a23e544a29a638610bdd6cb42a3e8b13030fd78334eba
```

  transaction_hash: String!
  account: String!
  value: BigInt!
  block_height: BigInt!
  block_timestamp: BigInt!
  daily_claim_summary: DailyClaimSummary!
}

type DailyClaimSummary @entity {
  id: ID! # this is the ISO string of the day e.g. '2023-03-26'
  total_claimed: BigInt!
}
```

::: warning Important
When you make any changes to the schema file, please ensure that you regenerate your types directory.
:::

SubQuery makes it easy and type-safe to work with your GraphQL entities, as well as smart contracts, events, transactions, and logs. SubQuery CLI will generate types from your project's GraphQL schema and any contract ABIs included in the data sources.

::: code-tabs
@tab:active yarn

```shell
yarn codegen
```

@tab npm

```shell
npm run-script codegen
```

:::

This will create a new directory (or update the existing one) `src/types` which contains generated entity classes for each type you have defined previously in `schema.graphql`. These classes provide type-safe entity loading, and read and write access to entity fields - see more about this process in [the GraphQL Schema](../../build/graphql.md). All entities can be imported from the following directory:

```ts

> import { Claim } from "../types/models/Claim";

Reviewer: ```suggestion
import { Claim, DailyClaimSummary  } from "../types";
```

> import { DailyClaimSummary } from "../types/models/DailyClaimSummary";

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

  // Read to see if there is an existing aggregation record
  let dailyClaimSummary = await DailyClaimSummary.get(id);
  if (!dailyClaimSummary) {
    // This is a new day! Create a new aggregation
    dailyClaimSummary = DailyClaimSummary.create({
      id,
      total_claimed: BigInt(0),
    });
  }
  return dailyClaimSummary;
}

export async function handleClaim(log: ClaimedLog): Promise<void> {
  logger.info(`New claimed log at block ${log.blockNumber}`);
  assert(log.args, "No log.args");

  // We make sure we have a daily claim aggregation to update
  const dailyClaimSummary = await checkGetDailyClaim(log.block.timestamp);

  const newClaim = Claim.create({
    id: log.args.index.toString(),
    account: log.args.account,
    transaction_hash: log.transactionHash,
    block_height: BigInt(log.blockNumber),
    block_timestamp: log.block.timestamp,
    value: log.args.amount.toBigInt(),
    daily_claim_summaryId: dailyClaimSummary.id,
  });

  // We update the daily aggregation
  dailyClaimSummary.total_claimed += newClaim.value;

  // Save data to the store
  await dailyClaimSummary.save();
  await newClaim.save();
}
```

> The `handleClaim` function receives a `log` parameter of type `ClaimedLog` which includes transaction log data in the payload. We extract this data, assign it to our newClaim object, and then save this to the store using the `.save()` function (_Note that SubQuery will automatically save this to the database_). We also call the checkGetDailyClaim function which checks if it is a new day and if it is, we create a new aggregation.

Reviewer: ```suggestion
The `handleClaim` function receives a `log` parameter of type `ClaimedLog` which includes transaction log data in the payload. We extract this data, assign it to our newClaim object, and then save this to the store using the `.save()` function (_Note that SubQuery will automatically save this to the database_). We also call the `checkGetDailyClaim` function to retrieve the existing day aggregation (and create a new one if we need to), and the update the `total_claimed` on it.
```
